src/vs/vscode.d.ts

```diff
@@ -7375,6 +7391,8 @@ declare module 'vscode' {
 	 */
 	export namespace languages {
 
+		export function setLanguageById(documentUri: Uri, languageId: string): Thenable<void>;
```

Comment:
Would be better to accept a `Document` here because we can only do that for existing documents. Then, I don't think that `set` is a good prefix because we aren't setting an instance property of the document but we re-create that document with a different language identifier. Maybe `changeLanguge` and make it return a document, e.g. `function changeLanguage(document: TextDocument, language: string): Thenable<TextDocument>`

Reply:
> we re-create that document with a different language identifier

Hmm, do you mean this code in extHostDocuments.ts?
```
    public $acceptModelModeChanged(uriComponents: UriComponents, oldModeId: string, newModeId: string): void {
        const uri = URI.revive(uriComponents);
        const strURL = uri.toString();
        let data = this._documentsAndEditors.getDocument(strURL);

        // Treat a mode change as a remove + add

        this._onDidRemoveDocument.fire(data.document);
        data._acceptLanguageId(newModeId);
        this._onDidAddDocument.fire(data.document);
    }
```
Because my experiments and the code shows that document doesn't get "recreated". The following condition in extHostDocumentData.ts prevents recreation of the document:
```
    get document(): vscode.TextDocument {
        if (!this._document) {
            const data = this;
            this._document = {
            ... 
``` 

But if I change `_acceptLanguageId` function there by adding the last line
```
    _acceptLanguageId(newLanguageId: string): void {
        ok(!this._isDisposed);
        this._languageId = newLanguageId;
        this._document = null;  // I've added this line!
    }
```
then the document gets recreated indeed.

Reply:
Well, we fire a `remove` and `add` event to underline at least the intention ;-) Unsure why we don't reset the document but I would leave it like that for now and tackle that independently.